tests/parser/syntax/test_unbalanced_return.py

```diff
@@ -0,0 +1,106 @@
+import pytest
+from pytest import (
+    raises,
+)
+
+from vyper import (
+    compiler,
+)
+from vyper.exceptions import (
+    StructureException,
+)
+
+fail_list = [
+    """
+@public
+def foo() -> int128:
+    pass
+    """,
+    """
+@public
+def foo() -> int128:
+    if False:
+        return 123
+    """,
+    """
+@public
+def test() -> int128:
+    if 1 == 1 :
+        return 1
+        if True:
+            return 0
+    else:
+        assert False
+    """,
+    """
+@private
+def valid_address(sender: address) -> bool:
+    selfdestruct(sender)
+    return True
+    """
+]
+
+
+@pytest.mark.parametrize('bad_code', fail_list)
+def test_missing_return(bad_code):
+    with raises(StructureException):
+        compiler.compile_code(bad_code)
+
+
+valid_list = [
+    """
+@public
+def foo() -> int128:
+    return 123
+    """,
+    """
+@public
+def foo() -> int128:
+    if True:
+        return 123
+    else:
+        raise "test"
+    """,
+    """
+@public
+def foo() -> int128:
+    if False:
+        return 123
+    else:
+        selfdestruct(msg.sender)
+    """,
+    """
+@public
+def foo() -> int128:
+    if False:
+        return 123
+    return 333
+    """,
+    """
+@public
+def test() -> int128:
+    if 1 == 1 :
+        return 1
+    else:
+        assert False
+        return 0
+    """,
+    """
+@public
+def test() -> int128:
+    x: bytes32
+    if False:
+        return 0
+    else:
+        x = sha3(x)
+        return 1
```

Comment:
This should fail despite the dead code block below it

Reply:
Meaning that this check should somehow have precedence over dead code

Reply:
Right, that's a dead code check.

Reply:
But it's in the "good" code list?

Reply:
Well yes, because all the exit points are balanced.

Reply:
Oh, so the return check does not account for anything after it, only that a balance exists? That makes sense why the dead code check us the mitigation

Reply:
Yes, but one can fix the tests for that scenario then ;)

Reply:
I think that's a bad strategy. Let's say someone else does that feature upgrade. They see this test case failing, so they move it to "bad code". Now we're missing the test of the happy case.

I think it is best to remove all potential dead code failures in here as out of scope. Should only be testing the return check logic.

Reply:
Like I said below, this check be added in this PR.